File: day_02/Day02P2.java
import java.io.*;
import java.util.*;

public class Day02P2 {
	private static boolean foo(List<Integer> nums, int start, boolean skip) {
		boolean inc = true, dec = true, inc_skip = skip, dec_skip = skip;
		int inc_last = start-1, dec_last = start-1;

		for (int i = start; i < nums.size(); ++i) {
			if (inc) {
				int diff = nums.get(i) - nums.get(inc_last);
				if (diff < 1 || diff > 3) {
					if (inc_skip) {
						inc_skip = false;
					} else {
						inc = false;
					}
				} else {
					inc_last = i;
				}
			}
			
			if (dec) {
				int diff = nums.get(dec_last) - nums.get(i);
				if (diff < 1 || diff > 3) {
					if (dec_skip) {
						dec_skip = false;
					} else {
						dec = false;
					}
				} else {
					dec_last = i;
				}
			}

			if (!inc && !dec) {
				break;
			}
		}
		return inc || dec;
	}
	public static void main(String[] args) {
		int ans = 0;

		try (BufferedReader reader = new BufferedReader(new FileReader("data.txt"))) {
			String line;
			
			while ((line = reader.readLine()) != null) {
				String[] parts = line.trim().split("\\s+");
				List<Integer> nums = Arrays.stream(parts)
					.map(Integer::parseInt)
					.toList();
				
				if (foo(nums, 1, true) || foo(nums, 2, false)) {
					ans++;
				}
			}

			System.out.println(ans);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}

File: day_02/day_02_p2.clj
(ns day_02_p2)

(defn check-diff [nums start skip_allowed]
  (let [inc_seq (atom true), dec_seq (atom true),
        inc_skip (atom skip_allowed), dec_skip (atom skip_allowed),
        inc_last (atom (dec start)), dec_last (atom (dec start))]
    (loop [i start]
      (if (= i (count nums)) true
        (do (when @inc_seq
              (let [diff (- (nth nums i) (nth nums @inc_last))]
                (if (or (< diff 1) (> diff 3))
                  (if @inc_skip (reset! inc_skip false)
                      (reset! inc_seq false))
                  (reset! inc_last i))))
            (when @dec_seq
              (let [diff (- (nth nums @dec_last) (nth nums i))]
                (if (or (< diff 1) (> diff 3))
                  (if @dec_skip (reset! dec_skip false)
                      (reset! dec_seq false))
                  (reset! dec_last i))))
          (if (not (or @inc_seq @dec_seq)) false (recur (inc i))))))))

(defn -main []
  (let [ans (atom 0)]
    (with-open [reader (clojure.java.io/reader "data.txt")]
      (doseq [line (line-seq reader)]
        (let [nums (map (fn [part] (Integer/parseInt part))
                        (clojure.string/split line #"\s+"))]
          (when (or (check-diff nums 1 true) (check-diff nums 2 false))
            (swap! ans inc)))))
    (println @ans)))
